src/Projection/ProcessEvents.php

```diff
@@ -0,0 +1,208 @@
+<?php
+/**
+ * This file is part of the prooph/mongodb-event-store.
+ * (c) 2018 prooph software GmbH <contact@prooph.de>
+ * (c) 2018 Sascha-Oliver Prolic <saschaprolic@googlemail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Prooph\EventStore\MongoDb\Projection;
+
+use Prooph\Common\Messaging\Message;
+use Prooph\EventStore\Exception;
+use Prooph\EventStore\Projection\ProjectionStatus;
+use Prooph\EventStore\StreamName;
+
+trait ProcessEvents
+{
+    private function processEvents(bool $keepRunning, bool $singleHandler): void
+    {
+        do {
+            $collectionNames = [];
+            $streamTimestampsStart = [];
+
+            foreach ($this->streamPositions as $streamName => $position) {
+                $collectionNames[$this->persistenceStrategy->generateCollectionName(new StreamName($streamName))] = $streamName;
+            }
+            // initialize stream to get new events during processing current events
+            $changeStream = $this->client->selectDatabase($this->database)->watch([
```

Comment:
What happens In a replay scenario that takes one hour or so? This could cause memory issues, or?

Reply:
Good point, we can add a [`batchSize`](https://docs.mongodb.com/manual/reference/method/db.collection.watch/) to avoid it.